carcara/src/ast/macros.rs

```diff
@@ -277,6 +277,14 @@ macro_rules! match_term_err {
 macro_rules! build_term {
     ($pool:expr, true) => { $pool.bool_true() };
     ($pool:expr, false) => { $pool.bool_false() };
+    // How to separate the (and c1 c2 c3) into another call to build_term
+    // so that an arbitrary term can sit there?
+    ($pool:expr, (choice (($z:literal $type:ty)) (and {$c1:expr} {$c2:expr} {$c3:expr}))) => {{
+        let int_sort = $pool.add($crate::ast::Term::Sort(Sort::Int));
+        let bindings = $crate::ast::BindingList(vec![($z.into(), int_sort.clone())]);
+        let body = build_term!($pool, (and {$c1} {$c2} {$c3}));
+        $pool.add(Term::Binder(Binder::Choice, bindings, body))
+    }};
```

Comment:
One way of doing it would be this:
```suggestion
    ($pool:expr, (choice (($z:literal $sort:ident)) $arg:tt)) => {{
        let sort = $pool.add($crate::ast::Term::Sort(Sort::$sort));
        let bindings = $crate::ast::BindingList(vec![($z.into(), sort)]);
        let body = build_term!($pool, $arg);
        $pool.add(Term::Binder(Binder::Choice, bindings, body))
    }};
```
Here we match a single arbitrary token tree as an argument (`$arg:tt`) and call `build_term!` recursively to build it.

Also, in the code you submitted, you are always using the `Sort::Int` sort instead of the sort provided to the macro. I changed it in this suggestion to take an identifier (`$sort:ident`), and construct the sort `Sort::$sort`. This will only work with the base variants of `Sort`, namely `Bool`, `Int`, `Real`, etc., but I think that's fine for now.

Reply:
Thank you! I was trying to bind the types as well, I'll do the changes now